src/clj/puppetlabs/services/jruby/jruby_puppet_core.clj

```diff
@@ -356,34 +339,139 @@
 
 (schema/defn ^:always-validate environment-class-info-entry
   :- EnvironmentClassInfoCacheEntry
-  "Create an environment class info entry"
-  ([]
-   (environment-class-info-entry nil))
-  ([tag :- (schema/maybe schema/Str)]
-   {:tag tag
-    :last-updated (System/currentTimeMillis)}))
+  "Create an environment class info entry.  The return value will have a nil
+  tag and cache-generation-id set to 1."
+  []
+  {:tag nil
+   :cache-generation-id 1})
+
+(schema/defn ^:always-validate inc-cache-generation-id-for-class-info-entry
+  :- EnvironmentClassInfoCacheEntry
+  "Return the supplied 'original-environment-class-info-entry' class info entry,
+  only updated with a cache-generation-id value that has been incremented.
+  If the supplied parameter is nil, the return value will have a tag set to
+  nil and cache-generation-id set to 1."
+  [original-environment-class-info-entry :-
+   (schema/maybe EnvironmentClassInfoCacheEntry)]
+  (if original-environment-class-info-entry
+    (update original-environment-class-info-entry
+            :cache-generation-id
+            #(if (= % Long/MAX_VALUE) Long/MIN_VALUE (inc %)))
```

Comment:
`9223372036854775807/(1000*60*60*24*365) = 292471208.68`... so it would take 292 million years to reach this if it were being hit every single millisecond for that amount of time. I would say either simplify this back to just `inc` or  add a comment about why this logic is there.